### PR TITLE
Use `--ssh=<host>[:<path>]` to run commands on remote servers

### PR DESCRIPTION
**Pull request by danielbachhuber**
See #2754

### CODE REVIEW

File: php/WP_CLI/Runner.php [hunk 1]
```diff
@@ -334,6 +334,44 @@ private function _run_command() {
...
+			$path = substr( $host, $key + 1 );
+			$host = substr( $host, 0, $key );
+		}
+
+		WP_CLI::debug( 'SSH host: ' . $host, 'bootstrap' );
+		WP_CLI::debug( 'SSH path: ' . $path, 'bootstrap' );
+
+		$is_tty = function_exists( 'posix_isatty' ) && posix_isatty( STDOUT );
+
+		$wp_binary = 'wp';
```
Comment (danielbachhuber) [inline php/WP_CLI/Runner.php]: @westonruter @jonathanbardo In situations where `wp` is defined by an alias or addition to `PATH` in a `.bash_profile`, have you found a way to load this?

Comment (miya0001) [inline php/WP_CLI/Runner.php]: `/usr/bin/env wp` ?
I haven't tested yet. 😄

Comment (jonathanbardo) [inline php/WP_CLI/Runner.php]: @danielbachhuber That's a valid scenario and I don't think we ever faced that problematic while we developed wp-cli-ssh.

Comment (danielbachhuber) [inline php/WP_CLI/Runner.php]: Hm, not quite:

```
local ➜  wp-cli git:(2754-ssh) ✗ wp --ssh=hb
/usr/bin/env: wp: No such file or directory
```

Comment (danielbachhuber) [inline php/WP_CLI/Runner.php]: > That's a valid scenario and I don't think we ever faced that problematic while we developed wp-cli-ssh.

Unfortunately, it's a problem with the first server I tested with :(

Comment (miya0001) [inline php/WP_CLI/Runner.php]: I can get $PATH via following command but alias doesn't work...
```
ssh user@example.com "source /etc/profile; echo \$PATH"
```

Comment (miya0001) [inline php/WP_CLI/Runner.php]: Ah it looks difficult.

* Place the `~/.ssh/environment` into local machine.
* Add `PermitUserEnvironment yes` in `/etc/ssh/sshd_config` on remote server.

http://stackoverflow.com/questions/216202/why-does-an-ssh-remote-command-get-fewer-environment-variables-then-when-run-man

Comment (westonruter) [inline php/WP_CLI/Runner.php]: @danielbachhuber This is why there was recent support for a `wpcli_command` config to handle situations where `wp` isn't on the `PATH`: https://github.com/xwp/wp-cli-ssh/pull/27

Comment (danielbachhuber) [inline php/WP_CLI/Runner.php]: > This is why there was recent support for a wpcli_command config to handle situations where wp isn't on the PATH: xwp/wp-cli-ssh#27

@westonruter Did you explore trying to get the data out of `.bash_profile` ? I'd prefer to not have to institute a workaround...

Comment (westonruter) [inline php/WP_CLI/Runner.php]: No, I didn't. Also, perhaps the PHAR isn't even aliased as `wp` on the target environment. I think that there will ultimately need to be a way to provide an alternate path.

**Comment by rmccue** [comment]
I'm not sure I see the difference between `wp --ssh=x ...` and `ssh x -c 'wp ...'`? Seems a little bit like reinventing stuff just for the sake of it to me.

**Comment by danielbachhuber** [comment]
> I'm not sure I see the difference between `wp --ssh=x` ... and `ssh x -c 'wp ...'`? Seems a little bit like reinventing stuff just for the sake of it to me.

This PR is the first part of #2039. Think `wp @all core update`

There's not much difference between `wp core download` and `wget https://wordpress.org/latest.zip`... except for the differences there are, and they are quite useful.

File: php/WP_CLI/Runner.php (hunk 1 above)
Comment (danielbachhuber) [inline php/WP_CLI/Runner.php]: > No, I didn't. Also, perhaps the PHAR isn't even aliased as wp on the target environment. I think that there will ultimately need to be a way to provide an alternate path.

I ended up creating a `WP_CLI_PRE_SSH_CMD` env variable, to offer a bit more flexibility:

```
WP_CLI::add_hook( 'before_ssh', function() {
	if ( $ssh = WP_CLI::get_runner()->config['ssh'] ) {
		list( $host, $path ) = explode( ':', $ssh );
		switch( $host ) {
			case 'hb':
				putenv( 'WP_CLI_SSH_PRE_CMD=export PATH=$HOME/bin:$PATH' );
				break;
		}
	}
});
```

**Pull request merged**